File: AOC_2022/Day 5/5.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <cstring>
#include <cctype>
#include <vector>
#include <regex>
#include <stack>
using namespace std;

int main()
{
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    string line;
    bool flag = false;
    vector<stack<char>> vec(9);
    while (getline(in, line))
    {
        if(!flag){
            if(line.size() == 0){
                for(int i = 0; i < 9;i++){
                    stack<char> tmp;
                    while(!vec[i].empty()){
                        tmp.push(vec[i].top());
                        vec[i].pop();
                    }
                    vec[i] = tmp;
                }
                flag = true;
            }
            else{
                for(int i = 0; i < line.size();i++){
                    if(line[i] == ' ') continue;
                    else if(line[i] == '['){
                        vec[i/4].push(line[i+1]);
                        i += 2;
                    }
                    else{
                        break;
                    }
                }
            }
        }
        else{
            regex rgx("move ((1[0-9])|([1-9])|(2[0-9])|(3[0-9])) from ([1-9]) to ([1-9])");
            smatch match;
            if (regex_search(line, match, rgx))
            {
                int count = 0, size, tostk, fromstk;
                for(size_t t = match.size()-1; t > 0;t--){
                    if(match[t].str().empty()){
                        continue;
                    }
                    else{
                        if(count == 0) tostk = stoi(match[t].str());
                        else if(count == 1) fromstk = stoi(match[t].str());
                        else if(count == 2) size = stoi(match[t].str());
                        else break;
                        count++;
                    }
                }
                //cout<<"move " <<size<<" from "<<fromstk<<" to "<<tostk<<"\n";
                while(size--){
                    vec[tostk-1].push(vec[fromstk-1].top());
                    vec[fromstk-1].pop();
                }
            }
        }
    }
    // for(auto it : vec){
    //     while(!it.empty()){
    //         cout << it.top() << " ";
    //         it.pop();
    //     }
    //     cout << "\n";
    // }
    for(auto it : vec){
        if(!it.empty()) cout<<it.top();
    }
    cout<<"\n";
    // part 2
    in.clear();
    in.seekg(0, ios::beg);
    flag = false;
    for(int i = 0; i < 9;i++){
        while(!vec[i].empty()){
            vec[i].pop();
        }
    }
    while (getline(in, line))
    {
        if(!flag){
            if(line.size() == 0){
                for(int i = 0; i < 9;i++){
                    stack<char> tmp;
                    while(!vec[i].empty()){
                        tmp.push(vec[i].top());
                        vec[i].pop();
                    }
                    vec[i] = tmp;
                }
                flag = true;
            }
            else{
                for(int i = 0; i < line.size();i++){
                    if(line[i] == ' ') continue;
                    else if(line[i] == '['){
                        vec[i/4].push(line[i+1]);
                        i += 2;
                    }
                    else{
                        break;
                    }
                }
            }
        }
        else{
            regex rgx("move ((1[0-9])|([1-9])|(2[0-9])|(3[0-9])) from ([1-9]) to ([1-9])");
            smatch match;
            if (regex_search(line, match, rgx))
            {
                int count = 0, size, tostk, fromstk;
                for(size_t t = match.size()-1; t > 0;t--){
                    if(match[t].str().empty()){
                        continue;
                    }
                    else{
                        if(count == 0) tostk = stoi(match[t].str());
                        else if(count == 1) fromstk = stoi(match[t].str());
                        else if(count == 2) size = stoi(match[t].str());
                        else break;
                        count++;
                    }
                }
                //cout<<"move " <<size<<" from "<<fromstk<<" to "<<tostk<<"\n";
                stack<char> temp;
                while(size--){
                    temp.push(vec[fromstk-1].top());
                    vec[fromstk-1].pop();
                }
                while(!temp.empty()){
                    //cout<<temp.top()<<" ";
                    vec[tostk-1].push(temp.top());
                    temp.pop();
                }
                //cout<<"\n";
            }
        }
    }
    // for(auto it : vec){
    //     while(!it.empty()){
    //         cout << it.top() << " ";
    //         it.pop();
    //     }
    //     cout << "\n";
    // }
    for(auto it : vec){
        if(!it.empty()) cout<<it.top();
    }
    cout<<"\n";
    return 0;
}

File: AOC_2022/Day 5/5.py
import re
file = open("input.txt","r")

lines = file.read().splitlines()

result = ""

flag = False

stks = [[] for x in range(9)]

#print(lines)

for line in lines:
    if not flag:
        if(len(line) == 0):
            for i in range(9):
                stks[i].reverse()
            flag = True
        else:
            i = 0
            while(i < len(line)):
                if line[i] ==  ' ':
                    i += 1
                    continue
                elif line[i] == '[':
                    stks[int(i/4)].append(line[i+1])
                    i += 2
                else: 
                    break
                i += 1
    else:
        matchregex = re.compile(r'move ((\d)|(1\d)|(2\d)|(3\d)) from (\d) to (\d)')
        mo = matchregex.search(line)
        size, fromStk, toStk = int(mo.group(1)), int(mo.group(6)), int(mo.group(7))
        #print(size, fromStk, toStk)
        while size:
            stks[toStk-1].append(stks[fromStk-1].pop())
            size -= 1

for i in range(9):
    if len(stks[i]): 
        result += stks[i][-1]

print("The Result for the 1st star is: ",result)

result = ""

stks = [[] for x in range(9)]

flag = False

for line in lines:
    if not flag:
        if(len(line) == 0):
            for i in range(9):
                stks[i].reverse()
            flag = True
        else:
            i = 0
            while(i < len(line)):
                if line[i] ==  ' ':
                    i += 1
                    continue
                elif line[i] == '[':
                    stks[int(i/4)].append(line[i+1])
                    i += 2
                else: 
                    break
                i += 1
    else:
        matchregex = re.compile(r'move ((\d)|(1\d)|(2\d)|(3\d)) from (\d) to (\d)')
        mo = matchregex.search(line)
        size, fromStk, toStk = int(mo.group(1)), int(mo.group(6)), int(mo.group(7))
        #print(size, fromStk, toStk)
        stks[toStk-1].extend(stks[fromStk-1][len(stks[fromStk-1])-size:])
        del stks[fromStk-1][len(stks[fromStk-1])-size:]

for i in range(9):
    if len(stks[i]): 
        result += stks[i][-1]
    

print("The Result for the 2nd star is: ",result)